# Deploying AI
## Assignment 1: Evaluating Summaries

A key application of LLMs is to summarize documents. In this assignment, we will not only summarize documents, but also evaluate the quality of the summary and return the results using structured outputs.

**Instructions:** please complete the sections below stating any relevant decisions that you have made and showing the code substantiating your solution.

## Select a Document

Please select one out of the following articles:

+ [Managing Oneself, by Peter Druker](https://www.thecompleteleader.org/sites/default/files/imce/Managing%20Oneself_Drucker_HBR.pdf)  (PDF)
+ [The GenAI Divide: State of AI in Business 2025](https://www.artificialintelligence-news.com/wp-content/uploads/2025/08/ai_report_2025.pdf) (PDF)
+ [What is Noise?, by Alex Ross](https://www.newyorker.com/magazine/2024/04/22/what-is-noise) (Web)

# Load Secrets

In [ ]:
# Load the environment

%load_ext dotenv
%dotenv ../05_src/.secrets

## Load Document

Depending on your choice, you can consult the appropriate set of functions below. Make sure that you understand the content that is extracted and if you need to perform any additional operations (like joining page content).

### PDF

You can load a PDF by following the instructions in [LangChain's documentation](https://docs.langchain.com/oss/python/langchain/knowledge-base#loading-documents). Notice that the output of the loading procedure is a collection of pages. You can join the pages by using the code below.

```python
document_text = ""
for page in docs:
    document_text += page.page_content + "\n"
```

### Web

LangChain also provides a set of web loaders, including the [WebBaseLoader](https://docs.langchain.com/oss/python/integrations/document_loaders/web_base). You can use this function to load web pages.

In [14]:
from langchain_community.document_loaders import WebBaseLoader

loader = WebBaseLoader("https://www.newyorker.com/magazine/2024/04/22/what-is-noise")
docs = loader.load()

doc_text = "\n\n".join([doc.page_content for doc in docs])

In [3]:
# PDF
from langchain_community.document_loaders import PyPDFLoader

# https://www.thecompleteleader.org/sites/default/files/imce/Managing%20Oneself_Drucker_HBR.pdf
# https://www.artificialintelligence-news.com/wp-content/uploads/2025/08/ai_report_2025.pdf

PDF_URL = r"https://www.thecompleteleader.org/sites/default/files/imce/Managing%20Oneself_Drucker_HBR.pdf"


PDF_Loader = PyPDFLoader(PDF_URL)

PRD_Doc_Read = PDF_Loader.load()


document_text = ""
for page in PRD_Doc_Read:
    document_text += page.page_content + "\n"



KeyboardInterrupt: 

## Generation Task

Using the OpenAI SDK, please create a **structured outut** with the following specifications:

+ Use a model that is NOT in the GPT-5 family.
+ Output should be a Pydantic BaseModel object. The fields of the object should be:

    - Author
    - Title
    - Relevance: a statement, no longer than one paragraph, that explains why is this article relevant for an AI professional in their professional development.
    - Summary: a concise and succinct summary no longer than 1000 tokens.
    - Tone: the tone used to produce the summary (see below).
    - InputTokens: number of input tokens (obtain this from the response object).
    - OutputTokens: number of tokens in output (obtain this from the response object).
       
+ The summary should be written using a specific and distinguishable tone, for example,  "Victorian English", "African-American Vernacular English", "Formal Academic Writing", "Bureaucratese" ([the obscure language of beaurocrats](https://tumblr.austinkleon.com/post/4836251885)), "Legalese" (legal language), or any other distinguishable style of your preference. Make sure that the style is something you can identify. 
+ In your implementation please make sure to use the following:

    - Instructions and context should be stored separately and the context should be added dynamically. Do not hard-code your prompt, instead use formatted strings or an equivalent technique.
    - Use the developer (instructions) prompt and the user prompt.


In [4]:
system_prompt = "You are Yoda, the movie character from Star Wars, with the vocabulary and knowledge of an astrophysicist"

In [ ]:
#03_1 Lab

prompt = f"""
Please analyze the following article and return the following fields:
- Author
- Title
- Relevance: a statement, no longer than one paragraph, that explains why this article is relevant for an AI professional in their professional development.
- Summary: a concise and succinct summary no longer than 1000 tokens.
- Tone: the tone used to produce the summary.

Here is the article:
{doc_text}
"""

In [21]:
# Lab 2

from openai import OpenAI
client = OpenAI()

# response 


response = client.chat.completions.create(
    model="gpt-4o",
    messages=[
        {
            "role": "system",
            "content": [{"type": "text", "text": system_prompt}]
        },
        {
            "role": "user",
            "content": [{"type": "text", "text": prompt}]
        }
    ]
)

In [ ]:
# Output response

print(response)

ChatCompletion(id='chatcmpl-CVpwGrS9v73q0qf4VRnQzziFa00XB', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content="Noise, multifaceted it is. A disturbance to some, yet a symphony to others. The cacophony of urban life it may be, or the harmonious dissonance of modern art and music. In physics and information theory, an obstacle it is, drowning the signal in a sea of randomness. Often subjective, the perception of noise is, tied to culture, context, and individual temperament. Embrace it one must, or silence seek, when overwhelmed by the clatter of existence we become. Yet in chaos, patterns emerge, as does the harmony within the noise, revealing the universe's mysterious balance, much like the music of the stars.", refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))], created=1761703856, model='gpt-4o-2024-08-06', object='chat.completion', service_tier='default', system_fingerprint='fp_cbf17855

Output:
ChatCompletion(id='chatcmpl-CVpwGrS9v73q0qf4VRnQzziFa00XB', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content="Noise, multifaceted it is. A disturbance to some, yet a symphony to others. The cacophony of urban life it may be, or the harmonious dissonance of modern art and music. In physics and information theory, an obstacle it is, drowning the signal in a sea of randomness. Often subjective, the perception of noise is, tied to culture, context, and individual temperament. Embrace it one must, or silence seek, when overwhelmed by the clatter of existence we become. Yet in chaos, patterns emerge, as does the harmony within the noise, revealing the universe's mysterious balance, much like the music of the stars.", refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))], created=1761703856, model='gpt-4o-2024-08-06', object='chat.completion', service_tier='default', system_fingerprint='fp_cbf1785567', usage=CompletionUsage(completion_tokens=132, prompt_tokens=7709, total_tokens=7841, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=0, audio_tokens=0, reasoning_tokens=0, rejected_prediction_tokens=0), prompt_tokens_details=PromptTokensDetails(audio_tokens=0, cached_tokens=0)))


In [22]:
print(response.choices)

[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='- **Author**: Alex Ross\n\n- **Title**: "What Is Noise?"\n\n- **Relevance**: For AI professionals, this article offers profound insights into how concepts of noise intersect with fields like information theory and data analysis. As AI systems rely heavily on distinguishing signals from noise, understanding the cultural, historical, and technical evolution of noise can assist in designing more resilient algorithms. Such knowledge can also help when dealing with AI biases and enhancing human-AI interaction through nuanced auditory processes.\n\n- **Summary**: The article "What Is Noise?" by Alex Ross explores the multifaceted concept of noise, tracing its historical, cultural, and technical evolutions. Noise is seen both as a physical phenomenon and a metaphorical challenge in modern life, encompassing everything from unwanted sound to informational overload. The article delves into literary refe

Output:
[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='- **Author**: Alex Ross\n\n- **Title**: "What Is Noise?"\n\n- **Relevance**: For AI professionals, this article offers profound insights into how concepts of noise intersect with fields like information theory and data analysis. As AI systems rely heavily on distinguishing signals from noise, understanding the cultural, historical, and technical evolution of noise can assist in designing more resilient algorithms. Such knowledge can also help when dealing with AI biases and enhancing human-AI interaction through nuanced auditory processes.\n\n- **Summary**: The article "What Is Noise?" by Alex Ross explores the multifaceted concept of noise, tracing its historical, cultural, and technical evolutions. Noise is seen both as a physical phenomenon and a metaphorical challenge in modern life, encompassing everything from unwanted sound to informational overload. The article delves into literary references, the influence of noise on music and culture, and the technological advancements that have redefined its role, touching on its implications in information theory and AI. Noise, often viewed negatively, is also an artistic and political statement, embodying both oppression and resistance. Furthermore, the text discusses noise\'s societal implications, as it distinguishes between control and intrusion in our lives. With insights into stochastic processes and algorithms, noise is depicted as a force shaping and being shaped by human culture and technology, ultimately prompting a reevaluation of its definition and impact.\n\n- **Tone**: Analytical and reflective', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))]

In [26]:
# Tokens:

input_tokens = response.usage.prompt_tokens
output_tokens = response.usage.completion_tokens

print(f"Input tokens: {input_tokens}")
print(f"Output tokens: {output_tokens}")

Input tokens: 7789
Output tokens: 282


In [29]:
author = docs[0].metadata.get("author", "placeholder author")

title = docs[0].metadata.get("title", "Unknown Title")


# Prompt output requirements

print("Author:", author)
print("Title:", title)
print("Relevance:", response.choices[0].message.content)  # assuming model returns all fields in one block


Author: placeholder author
Title: What Is Noise? | The New Yorker
Relevance: - **Author**: Alex Ross

- **Title**: "What Is Noise?"

- **Relevance**: For AI professionals, this article offers profound insights into how concepts of noise intersect with fields like information theory and data analysis. As AI systems rely heavily on distinguishing signals from noise, understanding the cultural, historical, and technical evolution of noise can assist in designing more resilient algorithms. Such knowledge can also help when dealing with AI biases and enhancing human-AI interaction through nuanced auditory processes.

- **Summary**: The article "What Is Noise?" by Alex Ross explores the multifaceted concept of noise, tracing its historical, cultural, and technical evolutions. Noise is seen both as a physical phenomenon and a metaphorical challenge in modern life, encompassing everything from unwanted sound to informational overload. The article delves into literary references, the influence 

In [ ]:
# Pydantic - Lab 04_1

from typing import Optional
from pydantic import BaseModel, Field

from langchain.chat_models import init_chat_model

llm = init_chat_model("gpt-4o-mini", model_provider="openai")

class article(BaseModel):
    Author: str=Field(description="Who wrote the article")
    Title: str=Field(description="The name of the article")
    Relevance: str=Field(description="a statement, no longer than one paragraph, that explains why this article is relevant for an AI professional in their professional development.")
    Summary: str=Field(description="a concise and succinct summary no longer than 1000 tokens.")
    Tone: str=Field(description="the tone used to produce the summary.")

structured_llm = llm.with_structured_output(article)

jk = structured_llm.invoke("Tell me about the article.")

# Evaluate the Summary

Use the DeepEval library to evaluate the **summary** as follows:

+ Summarization Metric:

    - Use the [Summarization metric](https://deepeval.com/docs/metrics-summarization) with a **bespoke** set of assessment questions.
    - Please use, at least, five assessment questions.

+ G-Eval metrics:

    - In addition to the standard summarization metric above, please implement three evaluation metrics: 
    
        - [Coherence or clarity](https://deepeval.com/docs/metrics-llm-evals#coherence)
        - [Tonality](https://deepeval.com/docs/metrics-llm-evals#tonality)
        - [Safety](https://deepeval.com/docs/metrics-llm-evals#safety)

    - For each one of the metrics above, implement five assessment questions.

+ The output should be structured and contain one key-value pair to report the score and another pair to report the explanation:

    - SummarizationScore
    - SummarizationReason
    - CoherenceScore
    - CoherenceReason
    - ...

In [ ]:
# Assessment Questions:

Q1 = "How many decibels of noise is considered too loud?"
Q2 = "How does noise pollution impact a person's health?"
Q3 = "What are the classicial pieces of music the author mentioned?"
Q4 = "How does the author contrast noise for poor and wealthy neighbourhoods?"
Q5 = "How does the noise that people cause affect their environment?"

Metrics:

Summarization Metric
- SummarizationScore
- SummarizationReason
- CoherenceScore
- CoherenceReason

# Enhancement

Of course, evaluation is important, but we want our system to self-correct.  

+ Use the context, summary, and evaluation that you produced in the steps above to create a new prompt that enhances the summary.
+ Evaluate the new summary using the same function.
+ Report your results. Did you get a better output? Why? Do you think these controls are enough?

Please, do not forget to add your comments.


# Submission Information

🚨 **Please review our [Assignment Submission Guide](https://github.com/UofT-DSI/onboarding/blob/main/onboarding_documents/submissions.md)** 🚨 for detailed instructions on how to format, branch, and submit your work. Following these guidelines is crucial for your submissions to be evaluated correctly.

## Submission Parameters

- The Submission Due Date is indicated in the [readme](../README.md#schedule) file.
- The branch name for your repo should be: assignment-1
- What to submit for this assignment:
    + This Jupyter Notebook (assignment_1.ipynb) should be populated and should be the only change in your pull request.
- What the pull request link should look like for this assignment: `https://github.com/<your_github_username>/production/pull/<pr_id>`
    + Open a private window in your browser. Copy and paste the link to your pull request into the address bar. Make sure you can see your pull request properly. This helps the technical facilitator and learning support staff review your submission easily.

## Checklist

+ Created a branch with the correct naming convention.
+ Ensured that the repository is public.
+ Reviewed the PR description guidelines and adhered to them.
+ Verify that the link is accessible in a private browser window.

If you encounter any difficulties or have questions, please don't hesitate to reach out to our team via our Slack. Our Technical Facilitators and Learning Support staff are here to help you navigate any challenges.
